src/Main.elm

```diff
@@ -1,10 +1,198 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser exposing (sandbox)
+import Data.Audience exposing (..)
+import Data.AudienceFolder exposing (..)
+import Dict exposing (Dict)
+import Dict.Extra as DE exposing (fromListDedupe)
+import Html exposing (Html, div, button, text)
+import Html.Events
+import Html.Attributes
+import Json.Decode as JD exposing (decodeString)
+import List as L exposing (map, drop, filter)
+import Maybe as M
 
 
-main : Html msg
-main =
-    Html.text "There will be app soon!"
+type alias Model = {
+    audiences : Dict Int (List Audience)
+    , audienceFolders :  Dict Int (List AudienceFolder)
+    , currentFolder : List Int
+    , audienceFilter : AudienceType }
+
+type Msg =
+    Up
+    | Down Int
+    | Filter AudienceType
+
+main : Program () Model Msg
+main = sandbox {
+    init = {audiences = audiences
+        , audienceFolders = audienceFolders
+        , currentFolder = []
+        , audienceFilter = Authored},
+    view = view,
+    update = update}
+
+view : Model -> Html Msg
+view model =
+    div [] [viewContents model]
+
+update : Msg -> Model -> Model
+update msg model =
+    case msg of
+        Up -> {model | currentFolder = L.drop 1 model.currentFolder}
+        Down dir -> {model | currentFolder = dir::model.currentFolder}
+        Filter at -> {model | audienceFilter = at, currentFolder = []}
+
+audiences : Dict Int (List Audience)
+audiences =
+    audiencesJSON
+    |> JD.decodeString decodeAudiences
+    |> Result.withDefault []
+    |> L.map (\aud -> (M.withDefault -1 aud.folder, [aud]))
+    |> DE.fromListDedupe List.append
+
+audienceFolders : Dict Int (List AudienceFolder)
+audienceFolders =
+    audienceFoldersJSON
+    |> JD.decodeString decodeAudienceFolders
+    |> Result.withDefault []
+    |> L.map (\audF -> (M.withDefault -1 audF.parent, [audF]))
+    |> DE.fromListDedupe List.append
+
+viewContents : Model -> Html Msg
+viewContents model =
+    let
+        currF = List.head model.currentFolder |> M.withDefault -1
+        contents =
+            if model.audienceFilter /= Shared then
+                List.append
+                (viewAudienceFolders model.audienceFolders currF)
+                (viewAudiences model.audienceFilter model.audiences currF)
+            else
+                viewAllAudiences
+                <| Dict.foldl (\parent children lst -> children++lst) [] model.audiences
+    in
+    Html.div (contentsStyle (currF /= -1))
+        [ maybeViewBackButton (currF /= -1 && model.audienceFilter /= Shared)
```

Comment:
I have to say I don't like the `-1` magic number much.
Elm allows us to use meaningful types like
```elm
type Position
  = Root -- instead of -1
  | Folder Int
```
So I'd prefer that

Reply:
Agree, this has potential to be buggy when you will somehow get something like `[ -1, 123, -1, 444 ]` so you can possibly ended up in folder structure without chance to going back. This is in general weak point of whole approach of using `List Int` for navigation. It's OK with current code, but code is usually changed over the time and as we know, if something can break it will break :)

Reply:
I agree. However, I knew the data in advance and database systems  (usually) start assigning IDs from 1. I was thinking about using `Maybe` but could not somehow get used to the idea, that I would be always `case`ing the type.
```
case head pathToRoot of 
    Just folder -> display folder's data
   Nothing -> displayRoot data
```
@Janiczek 's idea has the same "issue" with `case`ing, but at least it is clear that something is root and something is just Folder. However, a new case would arise => we would have to check whether the `pathToRoot` list is not empty.